modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/index/InlineUtils.java

```diff
@@ -0,0 +1,147 @@
+
+        boolean hasNullColumns = columns.stream().anyMatch(ColumnDescriptor::nullable);
+
+        // 2-byte size class covers most cases.
+        int sizeClass = 2;
+
+        if (columns.stream().allMatch(c -> c.type().spec().fixedLength())) {
+            int totalFixedColumnSize = columns.stream().mapToInt(c -> c.type().sizeInBytes()).sum();
+
+            sizeClass = Math.min(sizeClass, BinaryTupleCommon.flagsToEntrySize(BinaryTupleCommon.valueSizeToFlags(totalFixedColumnSize)));
+        }
+
+        int inlineSize = BinaryTupleCommon.HEADER_SIZE
+                + (hasNullColumns ? BinaryTupleCommon.nullMapSize(columns.size()) : 0)
+                + columns.size() * sizeClass;
+
+        for (int i = 0; i < columns.size() && inlineSize < MAX_BINARY_TUPLE_INLINE_SIZE; i++) {
+            inlineSize += inlineSize(columns.get(i).type(), maxVarlenInlineSize);
+        }
+
+        return Math.min(inlineSize, MAX_BINARY_TUPLE_INLINE_SIZE);
+    }
+
+    /**
+     * Calculates the inline size for the {@link BinaryTuple} that will be stored in the index {@link BplusInnerIo} and {@link BplusLeafIo}.
+     * Minimum number of items in a {@link BplusInnerIo} is {@code 2}, so that there is no significant performance drop.
+     *
+     * @param pageSize Page size in bytes.
+     * @param indexBplusIoHeaderSize Index {@link BplusInnerIo} and {@link BplusLeafIo} header size in bytes.
+     * @param indexDescriptor Index descriptor.
+     * @param maxVarlenInlineSize Maximum inline size of variable length column in bytes.
+     * @return Inline size in bytes, not more than the {@link #MAX_BINARY_TUPLE_INLINE_SIZE}.
+     */
+    public static int binaryTupleBplusIoInlineSize(
+            int pageSize,
+            int indexBplusIoHeaderSize,
+            IndexDescriptor indexDescriptor,
+            int maxVarlenInlineSize
+    ) {
+        int innerNodePayloadSize = pageSize - BplusInnerIo.HEADER_SIZE - PARTITIONLESS_LINK_SIZE_BYTES;
+        int leafNodePayloadSize = pageSize - BplusLeafIo.HEADER_SIZE;
+
+        // So that at least 2 items fit on the BplusInnerIo.
+        int maxInnerNodeItemSize = (innerNodePayloadSize / 2) - PARTITIONLESS_LINK_SIZE_BYTES;
+
+        int itemSize = Math.min(maxInnerNodeItemSize, binaryTupleInlineSize(indexDescriptor, maxVarlenInlineSize) + indexBplusIoHeaderSize);
+
+        // If there is a variable length column.
+        if (indexDescriptor.columns().stream().anyMatch(not(c -> c.type().spec().fixedLength()))) {
+            // Let's try to increase itemSize for BplusInnerIo.
+            int additionalInnerNodeItemSize = (innerNodePayloadSize % (itemSize + PARTITIONLESS_LINK_SIZE_BYTES))
+                    / (innerNodePayloadSize / (itemSize + PARTITIONLESS_LINK_SIZE_BYTES));
+
+            // Let's check if we can increase itemSize for BplusInnerIo so that the number of items for BplusLeafIo does not decrease.
+            if (additionalInnerNodeItemSize > 0
```

Comment:
This doesn't match what we discussed. I proposed choosing minimal value from two calculated deltas. But you either use a delta from inner page or a 0. Please fix it. Please also try to simplify this code by introducing some variables, like "innerPageMaxItems" or whatever.
By the way, why do you call them nodes here? Usually we refer to them as pages

Reply:
Fix it